### PR TITLE
C++: Proper MaD support for free functions

### PR DESCRIPTION
**Pull request by MathiasVP**
In https://github.com/github/codeql/pull/16791 we added syntax for specifying signatures in MaD that mention templates so that we could properly model `std` container member functions.

However, the PR didn't add the same support to non-member functions. This PR does that.

As a test, this PR also adds a MaD model for taint-flow through https://en.cppreference.com/w/cpp/utility/format/format. This has the nice bonus that it also exercises the code that was fixed in https://github.com/github/codeql/pull/17091

Commit-by-commit review recommended

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll [hunk 1]
```diff
@@ -776,13 +755,22 @@
   )
 }
 
+bindingset[name]
+pragma[inline_late]
+private predicate funcHasQualifiedName(Function func, string namespace, string name) {
+  exists(string nameWithoutArgs |
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll:761]: This actually highlighted a bug that's now fixed in 42ee501b96aba8e6f582ae3a2a07a79775455cd4 🎉

File: cpp/ql/lib/ext/std.format.model.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    data: # namespace, type, subtypes, name, signature, ext, input, output, kind, provenance
+      - ["std", "", False, "format<Args>", "(format_string,Args &&)", "", "Argument[*1]", "ReturnValue.Element[@]", "taint", "manual"]
+      - ["std", "", False, "format<Args>", "(format_string,Args &&)", "", "Argument[*2]", "ReturnValue.Element[@]", "taint", "manual"]
+      - ["std", "", False, "format<Args>", "(format_string,Args &&)", "", "Argument[*3]", "ReturnValue.Element[@]", "taint", "manual"]
+      - ["std", "", False, "format<Args>", "(format_string,Args &&)", "", "Argument[*4]", "ReturnValue.Element[@]", "taint", "manual"]
+      - ["std", "", False, "format<Args>", "(format_string,Args &&)", "", "Argument[*5]", "ReturnValue.Element[@]", "taint", "manual"]
+      - ["std", "", False, "format<Args>", "(format_string,Args &&)", "", "Argument[*6]", "ReturnValue.Element[@]", "taint", "manual"]
+      - ["std", "", False, "format<Args>", "(format_string,Args &&)", "", "Argument[*7]", "ReturnValue.Element[@]", "taint", "manual"]
+      - ["std", "", False, "format<Args>", "(format_string,Args &&)", "", "Argument[*8]", "ReturnValue.Element[@]", "taint", "manual"]
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/lib/ext/std.format.model.yml:13]: Also also, I think argument 0 should also taint the result (though hopefully in most cases argument 0 will be a constant).

**Review by geoffw0 — COMMENTED** [review]
LGTM apart from a few simple questions.  Thanks for addressing this!

File: cpp/ql/lib/ext/std.format.model.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/ext/std.format.model.yml:13]: > Why [@] on the ReturnValue.Element? This is a std::string so I'm not sure what we would be dereferencing.

Oops. That's a mistake, indeed. Will fix

[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/ext/std.format.model.yml:13]: Actually, I don't even think we should be putting an `Element` on the access path. We just want to taint the actual string

[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/lib/ext/std.format.model.yml]: Sounds good to me.  In practice the important thing is that we're consistent with other `std::string` models.  `StdStringConstructor`, for example, seems to think the (qualifier) object itself should be tainted.

**Comment by MathiasVP** [comment]
DCA is uneventful. I believe this is ready @geoffw0 🎉

**Comment by geoffw0** [comment]
I've replied on the discussion thread above.  I'm happy with the changes but unsure if we want `.Element` without changing other models to align.

**Comment by MathiasVP** [comment]
> I've replied on the discussion thread above. I'm happy with the changes but unsure if we want `.Element` without changing other models to align.

I can pull out the `Element` content, sure. Then we'll convert all `std::string` related models to `Element` content once-and-for-all at some later point to be sure we're consistent.

**Comment by geoffw0** [comment]
Hmm, when you put it like that, perhaps just writing up an issue to convert the other models is the way to do this.  It might be a bit too much to do in one PR anyway.

**Comment by MathiasVP** [comment]
Too late: 6d8a83fc1f161922c37e5ebe78b4202397f5fae3 😂 But yeah, feel free to write up an issue!

**Review by geoffw0 — APPROVED** [review]
:+1:

I'll write up the issue regarding `.Element`.

**Pull request merged**